[GStreamer] Clarify video sample passing semantics between video sink and player

philn opened the pull request:
#### b223a07ae64ec413b5d9289a56f255f08884cae8
<pre>
[GStreamer] Clarify video sample passing semantics between video sink and player
<a href="https://bugs.webkit.org/show_bug.cgi?id=254404">https://bugs.webkit.org/show_bug.cgi?id=254404</a>

Reviewed by Xabier Rodriguez-Calvar.

The player `triggerRepaint()` was reworked to make sample ownership passing more explicit.

* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoSinkCommon.cpp:
(webKitVideoSinkSetMediaPlayerPrivate):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::triggerRepaint):
(WebCore::MediaPlayerPrivateGStreamer::createVideoSink):
(WebCore::MediaPlayerPrivateGStreamer::repaintCallback): Deleted.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/262138@main">https://commits.webkit.org/262138@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b5220b579f63813385a1522147554cefd4e90f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/495 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/497 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/430 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/480 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/541 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/582 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/673 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/488 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/469 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/464 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/523 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/511 "2 api tests failed or timed out") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/453 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/500 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/425 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/470 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/455 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/473 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/420 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/464 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/159 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/463 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3556,14 +3556,14 @@ void MediaPlayerPrivateGStreamer::triggerRepaint(GstSample* sample)
         Locker sampleLocker { m_sampleMutex };
         shouldTriggerResize = !m_sample;
-        m_sample = sample;
+        m_sample = WTFMove(sample);
```
I wonder if you only need to lock when you replace the sample but not when you access it. You are changing it here but in the code you change below you use `m_sample` now without locking.

philn replied:
Lock added

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3986,7 +3981,10 @@ GstElement* MediaPlayerPrivateGStreamer::createVideoSink()
-        g_signal_connect_swapped(m_videoSink.get(), "repaint-requested", G_CALLBACK(repaintCallback), this);
+        g_signal_connect_swapped(m_videoSink.get(), "repaint-requested", G_CALLBACK(+[](MediaPlayerPrivateGStreamer* player, GstSample* sample) {
+            auto ownedSample = adoptGRef(sample);
+            player->triggerRepaint(WTFMove(ownedSample));
```
You can just `player->triggerRepaint(adoptGRef(sample));` and avoid the move and the local variable.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3556,15 +3556,19 @@ void MediaPlayerPrivateGStreamer::triggerRepaint(GstSample* sample)
-            return;
+        GRefPtr<GstCaps> caps;
+        {
+            Locker sampleLocker { m_sampleMutex };
```
I think it would be better to lock above and keep the lock until before `if (!m_canRenderingBeAccelerated) {`. Then there you can `unlockEarly()` . Then I think we would be good to go unless there is something I'm missing.

philn replied:
I would avoid invasive changes here, unless you found a possible deadlock case. Is it the case?

calvaris replied:
I didn't find a dead lock, but it is weird to see you're dumping a lock on something you're going to grab again microseconds later.

zdobersek commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3986,7 +3985,9 @@ GstElement* MediaPlayerPrivateGStreamer::createVideoSink()
         m_isUsingFallbackVideoSink = true;
         m_videoSink = webkitVideoSinkNew();
-        g_signal_connect_swapped(m_videoSink.get(), "repaint-requested", G_CALLBACK(repaintCallback), this);
+        g_signal_connect_swapped(m_videoSink.get(), "repaint-requested", G_CALLBACK(+[](MediaPlayerPrivateGStreamer* player, GstSample* sample) {
```
You have the `+` character embedded there.

philn replied:
Without it the build fails:

```
error: cannot cast from type '(lambda at /app/webkit/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGSt
reamer.cpp:3988:91)' to pointer type 'GCallback' (aka 'void (*)()')                                                                                                                                                  
        g_signal_connect_data ((m_videoSink.get()), ("repaint-requested"), (((GCallback) ([](MediaPlayerPrivateGStreamer* player, GstSample* sample) { player->triggerRepaint(adoptGRef(sample)); }))), (this), __nul
l, G_CONNECT_SWAPPED);                                                                                                                                                                                               
                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                 
1 error generated.                                                                                 
```

zdobersek replied:
:man_shrugging:

jyavenard replied:
I would pass the lambda as std::function instead to avoid ambiguity.

philn replied:
What is wrong with this anyway? We do  this in many places to convert lambdas to function pointers, expected by `G_CALLBACK()`. If this needs to change/improve, let's do it in another PR.

philn replied:
```
rg "G_CALLBACK\(\+" Source/ | wc -l                                                                                                                                                           
105
```

zdobersek replied:
There's nothing wrong, and std::function wouldn't work anyway. I forgot about the + operator being needed when wrapping into G_CALLBACK.

philn committed:
[GStreamer] Clarify video sample passing semantics between video sink and player
https://bugs.webkit.org/show_bug.cgi?id=254404

Reviewed by Xabier Rodriguez-Calvar.

The player `triggerRepaint()` was reworked to make sample ownership passing more explicit.

* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoSinkCommon.cpp:
(webKitVideoSinkSetMediaPlayerPrivate):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::triggerRepaint):
(WebCore::MediaPlayerPrivateGStreamer::createVideoSink):
(WebCore::MediaPlayerPrivateGStreamer::repaintCallback): Deleted.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

Canonical link: https://commits.webkit.org/262138@main
